lahja/trio/endpoint.py

```diff
@@ -0,0 +1,786 @@
+            return False
+        elif not self._running.is_set():
+            return False
+        elif self._finished.is_set():
+            return False
+        else:
+            return True
+
+    @property
+    def has_snappy_support(self) -> bool:
+        if self._has_snappy_support is None:
+            from lahja.snappy import is_snappy_available
+            self._has_snappy_support = is_snappy_available
+        return self._has_snappy_support
+
+    #
+    # Compression
+    #
+    def _compress_event(self, event: BaseEvent) -> Union[BaseEvent, bytes]:
+        if self.has_snappy_support:
+            import snappy
+            return cast(bytes, snappy.compress(pickle.dumps(event)))
+        else:
+            return event
+
+    #
+    # Establishing connections
+    #
+    def direct_connect(self, name: str, remote: RemoteEndpoint) -> None:
+        self.logger.debug(
+            "Endpoint %s direct connect to %s:%s",
+            self.name,
+            name,
+            remote,
+        )
+        if name in self._connected_endpoints:
+            raise ConnectionAttemptRejected(
+                f"Already connected to endpoint with name: {name}"
+            )
+        self._connected_endpoints[name] = remote
+
+    async def connect_to_endpoints(self, *endpoints: ConnectionConfig) -> None:
+        """
+        Connect to the given endpoints and await until all connections are established.
+        """
+        # TODO: need tests
+        async with trio.open_nursery as nursery:
+            for config in endpoints:
+                nursery.start_soon(self.connect_to_endpoint, config)
+
+    async def connect_to_endpoint(self, config: ConnectionConfig) -> None:
+        if not self.is_running:
+            raise Exception("TODO: cannot establish connections if endpoint isn't running")
+
+        if config.name in self._connected_endpoints:
+            raise ConnectionAttemptRejected(
+                f"Already connected to endpoint with name: {config.name}"
+            )
+        await wait_for_path(trio.Path(config.path))
+        remote = await RemoteEndpoint.connect_to(config)
+        self._connected_endpoints[config.name] = remote
+        # send the hello message to allow a reverse connection to be established.
+        await self.broadcast(_Hello(), BroadcastConfig(config.name))
+        # ensure we also read incoming messages from the other side of this connection
+        await self._reverse_connect_send_channel.send(remote)
+
+    def is_connected_to(self, endpoint_name: str) -> bool:
+        return endpoint_name in self._connected_endpoints
+
+    #
+    # Internal event processing logic
+    #
+    async def _process_item(self, item: BaseEvent, config: BroadcastConfig) -> None:
+        self.logger.debug(
+            "Endpoint %s processing item: %s  with config: %s",
+            self.name,
+            item,
+            config,
+        )
+        if item is TRANSPARENT_EVENT:
+            # TODO: why?
```

Comment:
I don't think we need this anymore. Back in the days where we where using `multiprocessing.Queue` and read from it using a `ThreadPoolExecutor`, I needed a way to break out of the `queue.get()` and basically followed the "poison pill" approach that was recommended here: https://stackoverflow.com/questions/52081033/how-to-shutdown-process-with-event-loop-and-executor/52081665#52081665